### PR TITLE
OCPQE-15774 - Periodic prunner for jobs more than 3 days

### PR DESCRIPTION
**Pull request by sgoveas**

### CODE REVIEW

**Review by aleskandro — COMMENTED** [review]
Hi @sgoveas 

I would like having two scripts:
1. Executes the check about clusters older than 3 days and call the second script to prune an older cluster, if any
2. Prune a cluster as in `prune_nodes "$CLUSTER"`.

The idea is that we should be able to do both manual clean up (using directly the second script), and allow the periodic pruning.

Finally, I see a lot of code here in a single function (that will become a script based on the above): ensure a good modularization in functions if possible and parametrization (for example, the hard coded INTERNAL_NET can allow changes from the launcher by using `INTERNAL_NET=${INTERNAL_NET:-192.168.90....}`, improve the in-code documentation to explain what we do here to future contributors... 

when ready for a review again, unhold this.

/hold

File: images/fcos-bastion-image/root/usr/bin/check_creation_time.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+# Check if any haproxy container is running for more than 3 days
+for i in $(docker ps -a --format "{{.Names}}" | grep "haproxy-")
+do
+  creation_ts=$(docker inspect --format="{{.Created}}" "$i")
+  create_ts=$(date -d "${creation_ts}" +%s)
+  now_ts=$(date +%s)
+  diff_in_ts=$((now_ts - create_ts))
+  time_in_days=$((diff_in_ts / 86400))
+  if [ "${time_in_days}" -ge 3 ]; then
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/check_creation_time.sh:11]:
```suggestion
  if [ "${diff_in_ts}" -ge 259200 ]; then # 3 Days * 86400 Seconds/Day
```

File: images/fcos-bastion-image/root/usr/bin/check_creation_time.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    CLUSTER=${i#haproxy-}
+    echo "<3>$CLUSTER is more than 3 days old, prunning started...."
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/check_creation_time.sh:13]:
```suggestion
    echo "<3>$CLUSTER is more than 3 days old, starting the pruning...."
```

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+conf_files=(/opt/dhcpd/root/etc/dnsmasq.conf
+            /opt/bind9_zones/zone
+            /opt/bind9_zones/internal_zone.rev
+            /opt/bind9_zones/external_zone.rev
+            /etc/vips_reserved
+            "${reserved_hosts}")
+
+readarray -t nodes < <(awk -F, -v cluster="${CLUSTER}" '$0 ~ cluster {print $12,$15,$16,$23}' "${reserved_hosts}")
+readarray -t node_ips < <(awk -F, -v cluster="${CLUSTER}" '$0 ~ cluster {print $2}' "${reserved_hosts}")
+readarray -t grub_macs < <(awk -F, -v cluster="${CLUSTER}" '$0 ~ cluster {print $1}' "${reserved_hosts}")
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:147]: there is a strong dependency on the number of the column here. Can we avoid it similarly to how we do in other code (see the scripts in openshift/release). The idea is that we should minimize the amount of changes required if we change the columns of the csv.

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 4]
```diff
@@ -0,0 +1,202 @@
+#!/bin/bash
+
+function haproxy_cleanup() {
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:3]: change the function such that it doesn't use global vars and try to do the same with the others

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    done
+  fi
+}
+
+function vlan_cleanup() {
+  if [ -s "${var_dir}"/vips.yaml ] && [ -s "${var_dir}"/hosts.yaml ]; then
+    ACTION=DELETE
+    VLAN_NAME=${CLUSTER}
+    VLAN_ID=$(grep "api_vip" "${var_dir}"/vips.yaml | awk -F. '{print $NF}')
+    INTERFACES=$(grep 'switch_port:' "${var_dir}"/hosts.yaml | awk '{print $NF}' | tr '\n' ',' | sed 's/,$//')
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:43]: use yq

**Review by aleskandro — COMMENTED** [review]
This starts to look right to me now...

However, I would:
1. Avoid massive use of global vars in favour of local vars or parameters passed to the functions.
2. use `yq` to avoid manual parsing errors when working with yaml files
3. define a clearer workflow in the "main" by reorganizing the way you execute loops... For example, the functions that execute `reboot_stuck_pdu` or `wait_for_node_status`, can be changed to be concurrent and called with a single host, such that you can execute a for loop and wait for the subprocs to finish in parallel.
```bash

for host in hosts:
  my_concurrent_safe_function my_arg my_arg2 &

wait
```

4. Sometimes you execute a check before running an action, to ensure that action is done against an existing asset (for example removing a container or removing a firewall rule). Other times, you don't. In my opinion, most of this script can avoid those checks and tolerate commands failures.... So you can try removing the container, even if it was already as well as you do with the deletion of the ovs ports. You can try removing the firewall rules of the bootstrap node by just running it for a given node, if it has `bootstrap` in its name.
5. Try to use only one source of truth for the hosts map... I'd suggest the hosts.yaml to be parsed as in https://github.com/openshift/release/blob/master/ci-operator/step-registry/baremetal/lab/pre/dns/baremetal-lab-pre-dns-commands.sh#L34-L46


/hold

File: images/fcos-bastion-image/root/usr/bin/check_creation_time.sh (hunk 1 above)
Comment (sgoveas) [reply images/fcos-bastion-image/root/usr/bin/check_creation_time.sh]: done

File: images/fcos-bastion-image/root/usr/bin/check_creation_time.sh (hunk 2 above)
Comment (sgoveas) [reply images/fcos-bastion-image/root/usr/bin/check_creation_time.sh]: done

File: images/fcos-bastion-image/root/usr/bin/prune_nodes (hunk 3 above)
Comment (sgoveas) [reply images/fcos-bastion-image/root/usr/bin/prune_nodes]: done

File: images/fcos-bastion-image/root/usr/bin/prune_nodes (hunk 4 above)
Comment (sgoveas) [reply images/fcos-bastion-image/root/usr/bin/prune_nodes:3]: done

File: images/fcos-bastion-image/root/usr/bin/prune_nodes (hunk 5 above)
Comment (sgoveas) [reply images/fcos-bastion-image/root/usr/bin/prune_nodes]: done

**Comment by sgoveas** [comment]
While cleaning up existing bastion host to test this script, found an issue with using hosts.yaml as source. A node used for bootstrap (deprovisioned) was included in the cleanup , but that was being used for a new job, maybe we should cross check with the hosts_pool_reserved file.

**Comment by aleskandro** [comment]
> While cleaning up existing bastion host to test this script, found an issue with using hosts.yaml as source. A node used for bootstrap (deprovisioned) was included in the cleanup , but that was being used for a new job, maybe we should cross check with the hosts_pool_reserved file.

This means there could be a bug in the current automation. We need to fix it (in the prow install steps for upi, when destroying the bootrap node) rather than doing a cross-check here.

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  done
+}
+
+############
+ ## Main ##
+############
+
+if [ "$#" -ne 1 ]; then
+    echo "<3>Usage: $0 <cluster-name>"
+    return 1
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:149]:
```suggestion
    exit 1
```

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  ## Remove port forwarding from bastion host to bootstrap host
+  ## Remove bootstrap host access limit to internet if in disconnected network
+  ## Remove cluster nodes access limit to internet if in disconnected network
+  firewall_cleanup "${CLUSTER}"
+
+  # Delete VLAN on Juniper
+  vlan_cleanup "${CLUSTER}"
+
+  # Reboot any stuck node pdu
+  exec_on_nodes reboot_stuck_pdu &
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:174]: Do not start this in the background. Let's do it within the loop.

```suggestion
  exec_on_nodes reboot_stuck_pdu 
```

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  for bmhost in $(yq e -o=j -I=0 '.[]' "${SHARED_DIR}/hosts.yaml"); do
+    # shellcheck disable=SC1090
+    . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
+    if [ ${#bmc_address} -eq 0 ] || [ ${#bmc_user} -eq 0 ] || [ ${#bmc_pass} -eq 0 ] || [ ${#name} -eq 0 ]; then
+      echo "Error while unmarshalling hosts entries"
+      exit 1
+    fi
+
+    # Execute the action
+    "${action}" "${bmc_address}" "${bmc_user}" "${bmc_pass}" "${name}" "${pdu_uri:-}" "${params[@]}"
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:138]:
```suggestion
    "${action}" "${bmc_address}" "${bmc_user}" "${bmc_pass}" "${name}" "${pdu_uri:-}" "${params[@]}" &
```

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+  done
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:140]:
```suggestion
  done
  wait
```

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+  exec_on_nodes reboot_stuck_pdu &
+
+  # Wait if a node pdu was rebooted
+  if [ "${pdu_reboot}" = "true" ]; then
+    echo "<4>Waiting for 2 min before checking node response..." && sleep 120
+    exec_on_nodes wait_for_node_status 'for pdu reboot' "' on| off'"
+  fi
+
+  # Wipe disk and power off
+  exec_on_nodes wipe_disk &
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:183]:
```suggestion
  exec_on_nodes wipe_disk
```

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+  vlan_cleanup "${CLUSTER}"
+
+  # Reboot any stuck node pdu
+  exec_on_nodes reboot_stuck_pdu &
+
+  # Wait if a node pdu was rebooted
+  if [ "${pdu_reboot}" = "true" ]; then
+    echo "<4>Waiting for 2 min before checking node response..." && sleep 120
+    exec_on_nodes wait_for_node_status 'for pdu reboot' "' on| off'"
+  fi
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:180]: this check (wait_for_node_status..._) go in the reboot_stuck_pdu function, on a per-node basis.

File: images/fcos-bastion-image/root/usr/bin/prune_nodes [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+  # Wipe disk and power off
+  exec_on_nodes wipe_disk &
+
+  # Wait to confirm nodes have powered off
+  echo "<4>Waiting for 2 min before checking node power off status..." && sleep 120
+  exec_on_nodes wait_for_node_status 'to power off' 'Power is off'
```
Comment (aleskandro) [inline images/fcos-bastion-image/root/usr/bin/prune_nodes:187]: Same as above, let's check this on a per-node basis in the wait_for_node_status function.

If needed, add additional parameters to distinguish the pdu-reboot from the ipmi one.

**Review by aleskandro — COMMENTED** [review]
Hi @sgoveas everything is almost good. Just some requests to adjust the concurrent functions

**Pull request merged**